pkg/service/service.go

```diff
@@ -15,108 +15,172 @@ import (
+	"github.com/gorilla/mux"
 
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
 	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
 	"github.com/google/uuid"
 )
 
 var (
-	svc *Service
+	sdk *DeviceServiceSDK
 )
 
-// A Service listens for requests and routes them to the right command
-type Service struct {
-	svcInfo     *common.ServiceInfo
-	asyncCh     chan *dsModels.AsyncValues
-	deviceCh    chan []dsModels.DiscoveredDevice
-	startTime   time.Time
-	controller  controller.RestController
-	initiazlied bool
+type DeviceServiceSDK struct {
+	ServiceName    string
+	LoggingClient  logger.LoggingClient
+	registryClient registry.Client
+	edgexClients   clients.EdgeXClient
+	controller     controller.RestController
+	config         *common.ConfigurationStruct
+	svcInfo        *common.ServiceInfo
+	deviceService  contract.DeviceService
+	driver         dsModels.ProtocolDriver
+	discovery      dsModels.ProtocolDiscovery
+	asyncCh        chan *dsModels.AsyncValues
+	deviceCh       chan []dsModels.DiscoveredDevice
+	initialized    bool
+}
+
+// TODO: remove common.xxx = xxx assignment after refactor are done.
+func (s *DeviceServiceSDK) Initialize(serviceName, serviceVersion string, proto interface{}) {
+	if serviceName == "" {
+		_, _ = fmt.Fprintf(os.Stderr, "Please specify device service name")
+		os.Exit(1)
+	}
+	common.ServiceName = serviceName
+	s.ServiceName = serviceName
+
+	if serviceVersion == "" {
+		_, _ = fmt.Fprintf(os.Stderr, "Please specify device service version")
+		os.Exit(1)
+	}
+	common.ServiceVersion = serviceVersion
+
+	if driver, ok := proto.(dsModels.ProtocolDriver); ok {
+		common.Driver = driver
+		s.driver = driver
+	} else {
+		_, _ = fmt.Fprintf(os.Stderr, "Please implement and specify the protocoldriver")
+		os.Exit(1)
+	}
+
+	if discovery, ok := proto.(dsModels.ProtocolDiscovery); ok {
+		common.Discovery = discovery
+		s.discovery = discovery
+	} else {
+		s.discovery = nil
+	}
+
+	s.config = &common.ConfigurationStruct{}
+}
+
+func (s *DeviceServiceSDK) UpdateFromContainer(r *mux.Router, dic *di.Container) {
+	s.LoggingClient = bootstrapContainer.LoggingClientFrom(dic.Get)
+	s.registryClient = bootstrapContainer.RegistryFrom(dic.Get)
+	s.edgexClients.GeneralClient = container.MetadataGeneralClientFrom(dic.Get)
+	s.edgexClients.DeviceClient = container.MetadataDeviceClientFrom(dic.Get)
+	s.edgexClients.DeviceServiceClient = container.MetadataDeviceServiceClientFrom(dic.Get)
+	s.edgexClients.DeviceProfileClient = container.MetadataDeviceProfileClientFrom(dic.Get)
+	s.edgexClients.AddressableClient = container.MetadataAddressableClientFrom(dic.Get)
+	s.edgexClients.ProvisionWatcherClient = container.MetadataProvisionWatcherClientFrom(dic.Get)
+	s.edgexClients.EventClient = container.CoredataEventClientFrom(dic.Get)
+	s.edgexClients.ValueDescriptorClient = container.CoredataValueDescriptorClientFrom(dic.Get)
+
+	s.svcInfo = &container.ConfigurationFrom(dic.Get).Service
```

Comment:
Why not setting `s.config` from container?

Reply:
udpated.